@atproto/api v0.14.0 release notes (#307)
bluesky-social/bsky-docs: blog/2025-01-13-typescript-sdk-release.md

---
slug: api-v0-14-0-release-notes
title: '@atproto/api v0.14.0 release notes'
tags: ['updates', 'guide']
---

> Today we are excited to announce the availability of version 0.14 or our TypeScript SDK on [NPM](https://www.npmjs.com/package/@atproto/api).

Reviewer: ```suggestion
Today we are excited to announce the availability of version 0.14 or our TypeScript SDK on [npm](https://www.npmjs.com/package/@atproto/api).
```

reallllly nit: npm styleguide is to use lowercase lol

const myPost: BlueskyPost = {
  text: 'Hey, check this out!',
  createdAt: '2021-09-01T12:34:56Z',
  embed: {
    $type: 'app.bsky.embed.video',
    video: 43, // This is invalid, but TypeScript does not complain
  },
}
```

We have fixed these issues by making the `$type` property in the generated interfaces explicit. The `app.bsky.embed.video` interface now looks like this:

```typescript
export interface Main {
  $type?: $Type<'app.bsky.embed.video', 'main'>
  video: BlobRef
  captions?: Caption[]
  alt?: string
  aspectRatio?: AppBskyEmbedDefs.AspectRatio
}
```

Notice how the `$type` property is defined as optional (`?:`) here. This is due to the fact that lexicons can define schemas that can be referenced from other places than open unions. In those places, there might not be any ambiguity as to how the data should be interpreted. For example, an embed that represents a "Record With Media" has a `record` property that expects an `app.bsky.embed.record` object:

```typescript
export interface Main {
  $type?: $Type<'app.bsky.embed.recordWithMedia', 'main'>
  record: AppBskyEmbedRecord.Main // Also used in post's `embed` property
  media: /* omitted */
}
```

> Since there is no ambiguity as to the type of the data here, making the `$type` property required would cause unnecessary bloat. Making the `$type` property optional allows to declare a "Record With Media" as follows:

Reviewer 2: ```suggestion
Since there is no ambiguity as to the type of the data here, making the `$type` property required would cause unnecessary bloat. Making the `$type` property optional allows declaring a "Record With Media" as follows:
```
I think this would read a bit clearer

> Since there is no ambiguity as to the type of the data here, making the `$type` property required would cause a huge un-necessary bloat. Making the `$type` property optional allows to declare a "Record With Media" as follows:

Reviewer: ```suggestion
Since there is no ambiguity as to the type of the data here, making the `$type` property required would cause unnecessary bloat. Making the `$type` property optional allows to declare a "Record With Media" as follows:
```

```typescript
const recordWithMedia: RecordWithMedia = {
  $type: 'app.bsky.embed.recordWithMedia',
  record: {
    // $type is not needed here
    record: {
      /* omitted */
    },
  },
  media: {
    /* omitted */
  },
}
```

Because the `$type` property of objects is required in some contexts while optional in others, we introduced a new utility type to make it required when needed. The `$Typed` utility allows to mark an interface’s `$type` property non optional in contexts where it is required:

```typescript
export type $Typed<V> = V & { $type: string }
```

> That utility allows to mark an interface’s `$type` property non optional. The `embed` property of posts is now defined as follows:

Reviewer 2: ```suggestion
The `embed` property of posts is now defined as follows:
```
the first sentence here seems to be repeating what was just said, so I think it can be removed?

```typescript
export interface Record {
  // ...
  embed?:
    | $Typed<AppBskyEmbedImages.Main>
    | $Typed<AppBskyEmbedVideo.Main>
    | $Typed<AppBskyEmbedExternal.Main>
    | $Typed<AppBskyEmbedRecord.Main>
    | $Typed<AppBskyEmbedRecordWithMedia.Main>
    | { $type: string }
}
```

> In addition to preventing the _creation_ of invalid data as seen at the beginning of this section, this change also allows to properly discriminate types when _accessing_ the data. For example, one can now do:

Reviewer 2: ```suggestion
In addition to preventing the _creation_ of invalid data as seen at the beginning of this section, this change also allows properly discriminating types when _accessing_ the data. For example, one can now do:
```
grammar nit

```tsx
import { AppBskyFeedPost } from '@atproto/api'

// Aliased for clarity
type BlueskyPost = AppBskyFeedPost.Main

// Say we got some random post somehow (typically via an api call)
declare const post: BlueskyPost

// And we want to know what kind of embed it contains
const { embed } = post

// We can now use the `$type` property to disambiguate
if (embed?.$type === 'app.bsky.embed.images') {
  // The `embed` variable is fully typed as `$Typed<AppBskyEmbedImages.Main>` here !
}
```

### `is*` utility methods

The example above shows how data can be discriminated based on the `$type` property. There are, however, several disadvantages to relying on string comparison for discriminating data types:

> - Having to use inline strings yields a lot of code, hurting readability and bundle size.

Reviewer 2: ```suggestion
- Having to use inline strings yields a lot of code, hurting readability and increasing bundle size.
```
this would read slightly clearer I think

- In particular instances, the `$type` property can actually have two values to describe the same lexicon. An "images" embed, for example, can use both `app.bsky.embed.images` and `app.bsky.embed.images#main` as `$type`. This makes the previous point even worse.

> In order to alleviate these issues, the SDK provides type checking predicate functions. In their previous implementation, the `is*` utilities where defined as follows:

Reviewer: ```suggestion
In order to alleviate these issues, the SDK provides type checking predicate functions. In their previous implementation, the `is*` utilities were defined as follows:
```

```typescript
import { AppBskyEmbedImages } from '@atproto/api'

// Alias, for clarity
const isImages = AppBskyEmbedImages.isMain

// Get an invalid embed somehow
const invalidEmbed = {
  $type: 'app.bsky.embed.images',
  // notice how the `images` property is missing here
}

// This predicate function only checks the value of the `$type` property, making the condition "true" here
if (isImages(invalidEmbed)) {
  // No TypeScript error, BUT causes a runtime error because there is no "images" property !
  console.log('First image:', invalidEmbed.images[0])
}
```

The root of the issue here is that the `is*` utility methods perform type casting of objects solely based on the value of their `$type` property. There were basically two ways of fixing this issue:

1. Alter the implementation to actually validate the object's structure. This is a non-breaking change that has a negative impact on performance.
2. Alter the function signature to describe what the function actually does. This is a breaking change because TypeScript would start (rightfully) returning lots of errors in places where these functions are used.

> Because this release introduces other breaking changes, and because adapting our own code base to this change showed it made more sense, we decided to adopt the latter option.

Reviewer: ```suggestion
Because this release introduces other breaking changes, and because adapting our own codebase to this change showed it made more sense, we decided to adopt the latter option.
```

> In lots of cases where data needs to be discriminated, this change in the signature of the `is*` function will actually not cause any issue when upgrading the version of the SDK. This is the case for example when working with data obtained from the API. Because an API is a "contract" between a server and a client, the data returned by the server is "guaranteed" to be valid. In these cases, the `is*` utility methods provide a convenient way to discriminate between valid values.

Reviewer: This note that we guarantee server response shape should be more prominent imo. Do we have admonitions set up here? If nothing else, just bolded.

Devs migrating to this _need_ to know this, otherwise they're going to think they need to use the `isValid*` method variants.

> In lots of cases where data needs to be discriminated, this change in the signature of the `is*` function will actually not cause any issue when upgrading the version of the SDK. This is the case for example when working with data obtained from the API. Because an API is a "contract" between a server and a client, the data returned by the server is "guaranteed" to be valid. In these cases, the `is*` utility methods provide a convenient way to discriminate between valid values.

Reviewer 2: ```suggestion
In lots of cases where data needs to be discriminated, this change in the signature of the `is*` function will actually not cause any issue when upgrading the version of the SDK. For example, this is the case when working with data obtained from the API. Because an API is a "contract" between a server and a client, the data returned by the server is "guaranteed" to be valid. In these cases, the `is*` utility methods provide a convenient way to discriminate between valid values.
```
I think moving `for example` to the beginning would read more clearly

```typescript
import { AppBskyEmbedImages } from '@atproto/api'

// Aliased for clarity
const isImages = AppBskyEmbedImages.isMain

// Get a post from the API (the API's contract guarantees the validity of the data)
declare const post: BlueskyPost

// The `is*` utilities are an efficient way to discriminate **valid** data based on their `$type`
if (isImages(post.embed)) {
  // `post.embed` is fully typed as `$Typed<AppBskyEmbedImages.Main>` here !
}
```

For other cases, when the data's validity is not known at dev time, we added new `isValid*` utility methods allowing to ensure that a value properly satisfies the interface.

```typescript
import { AppBskyEmbedImages } from '@atproto/api'

// Aliased for clarity
const Images = AppBskyEmbedImages.Main
const isValidImages = AppBskyEmbedImages.isValidMain

// Get an embed with unknown validity somehow
declare const embed: unknown

> // The following condition will be true if, and only if, the value matches the `Image` interface

Reviewer: Nit: lot of these comments throughout this doc scroll off the screen. Might be good to do a multi-line comment so that readers don't have to scroll horizontally to read the comments.

if (isValidImages(embed)) {
  // `embed` is of type `Images` here
}
```

> These method perform data validation, making them somewhat slower than the `is*` utility methods. They can however be used in place of the `is*` utilities when migrating to this new version of the SDK.

Reviewer 2: ```suggestion
These methods perform data validation, making them somewhat slower than the `is*` utility methods. They can, however, be used in place of the `is*` utilities when migrating to this new version of the SDK.
```
a typo and a style nit

## `validate*` utility methods

As part of this update, the signature of the `validate*` utility methods was updated to properly describe the type of the `value` in case of success:

```typescript
import { AppBskyEmbedImages } from '@atproto/api'

// Aliased for clarity
const Images = AppBskyEmbedImages.Main
const validateImages = AppBskyEmbedImages.validateMain

// Get some date somehow
declare const data: unknown

// Validate the data against a particular schema (images here)
const result = validateImages(data)

if (result.success) {
  // The `value` property was previously typed as `unknown` and is now properly typed as `Image`
  const images = result.value
}
```

## Removal of the `[x: string]` index signature

> Another property of Atproto being an "open protocol" is the fact that objects are allowed to contain additional, unspecified, properties (though this should be done with caution to avoid incompatibility with properties added in the future). This used to be represented in the type system using a `[k: string]: unknown` index signature in generated interfaces. This is how the video embed was represented:

Reviewer 2: ```suggestion
Another property of Atproto being an "open protocol" is the fact that objects are allowed to contain additional, unspecified, properties (although this should be done with caution to avoid incompatibility with properties that are added in the future). This used to be represented in the type system using a `[k: string]: unknown` index signature in generated interfaces. This is how the video embed was represented:
```
slight suggested wording change
